Feature: adds configurable query parameters to tile endpoints

bemyak opened the pull request.

dechristopher committed:
fix: minor warnings

dechristopher committed:
feat: add Query Parameters to Map config

fix: small typos and doc cleanup in provider/postgis
feat: implemented Query Parameters into postgres provider
feat: implemented default parameter values
feat: revamp param specification in config

roelarents committed:
handle gpkg GEOMETRY as unknown geometry

iwpnd committed:
feat: structured logs with uber/zap

As discussed in #831 I added the option to use structured logs with uber/zap
while falling back to the currently implemented default if --logger
is not specifically used. Log levels conform to the current implementation
with the exception of log level TRACE that is not supported by zap.
I use Debug as a fall back here.

dependabot[bot] committed:
Bump eventsource from 1.1.0 to 1.1.1 in /ui

Bumps [eventsource](https://github.com/EventSource/eventsource) from 1.1.0 to 1.1.1.
- [Release notes](https://github.com/EventSource/eventsource/releases)
- [Changelog](https://github.com/EventSource/eventsource/blob/master/HISTORY.md)
- [Commits](https://github.com/EventSource/eventsource/compare/v1.1.0...v1.1.1)

---
updated-dependencies:
- dependency-name: eventsource
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

iwpnd committed:
test: add a ttools helper for easier local testing

Previously a local test required to setup the local development
environment and passing a bunch of environment variables prior to
running `go test ./...`. Now we allow for environment variables to be
passed, but also default back to variables we used across tests and
pipelines as default, making testing a little more comfy.

test: add additional tests

fix: use os.LookupEnv instead

docs: add comment

bemyak committed:
refactor: thread query parameters all the way to provider

kszafran commented on cmd/tegola/cmd/root.go:
```diff
@@ -64,11 +64,11 @@ var RootCmd = &cobra.Command{
 	Short: "tegola is a vector tile server",
 	Long: fmt.Sprintf(`tegola is a vector tile server
-Version: %v`, build.Version),
+		Version: %v`, build.Version),
```
Wouldn't it add literal tabs to the string?

bemyak committed:
feat: replaceParams with tests

kszafran commented on config/config.go:
```diff
@@ -16,6 +18,64 @@ import (
+
+// IsReservedToken returns true if the specified token is reserved
+func IsReservedToken(token string) bool {
+	for _, t := range ReservedTokens {
```
You could instead make `ReservedTokens` a `map[string]bool`.

kszafran commented on config/config.go:
```diff
@@ -52,12 +112,86 @@ type Webserver struct {
+		return nil
+	}
+
+	var usedNames, usedTokens []string
```
These could also be maps to avoid iterating over them.

kszafran commented on config/config.go:
```diff
@@ -101,6 +235,30 @@ func (ml MapLayer) GetName() (string, error) {
+	if len(param.SQL) > 0 {
+		sql = param.SQL
+	}
+	param.SQL = sql
```
🤔 Isn't this just a convoluted version of:
```go
if len(param.SQL) == 0 {
    param.SQL = "?"
}
```
?

kszafran commented on config/config.go:
```diff
@@ -101,6 +235,30 @@ func (ml MapLayer) GetName() (string, error) {
+	Type  string `toml:"type"`
+	SQL   string `toml:"sql"`
+	// DefaultSQL replaces SQL if param wasn't passed. Either default_sql or
+	// default_value can be specified
```
What happens if I don't specify neither `default_value` nor `default_sql`?

bemyak replied:
The parameter will be marked as `required` and the client will get 400 if it is missing

kszafran commented on config/config_test.go:
```diff
@@ -169,7 +169,26 @@ func TestParse(t *testing.T) {
+
+					[[maps.params]]
+					name = "param1"
+					token = "!param1!"
```
Should we require tokens to be uppercase?

bemyak replied:
We make them uppercase during parsing, but maybe we should fail validation in such cases?

kszafran replied:
Yeah, I'd do that for the sake of clarity.

bemyak replied:
Hm, this turned out to be a bit problematic. Currently, all reserved tokens are allowed to be lowercase and will be transformed to uppercase internally. Having a different approach for parameter tokens introduces discrepancy in how they are parsed and handled.

kszafran replied:
Oh. In this case, let's be consistent with the current behavior.

kszafran commented on provider/postgis/util.go:
```diff
@@ -169,30 +155,63 @@ func replaceTokens(sql string, lyr *Layer, tile provider.Tile, withBuffer bool)
+		for _, c := range param.SQL {
+			if c == '?' {
+				*args = append(*args, param.Value)
+				paramSQL = paramSQL + "$" + fmt.Sprint(len(*args))
```
Or:
```suggestion
				paramSQL += paramSQL + fmt.Sprintf("$%d", len(*args)))
```

bemyak replied:
If I'm reading your suggestion right, it will desugar into:
```go
paramSQL = paramSQL + paramSQL + fmt.Sprintf("$%d", len(*args)))
```

Guess the second `paramSQL` should not be there ;)

kszafran replied:
Ah, yeah, it was supposed to be:
```go
paramSQL += fmt.Sprintf("$%d", len(*args)))
```
😅

kszafran commented on provider/postgis/util.go:
```diff
@@ -169,30 +155,63 @@ func replaceTokens(sql string, lyr *Layer, tile provider.Tile, withBuffer bool)
+		paramSQL := ""
+		for _, c := range param.SQL {
+			if c == '?' {
+				*args = append(*args, param.Value)
```
Hm... I guess this way you'll end up with e.g.:
```sql
? = "one" OR ? = "two"
```
even though this should work, too:
```sql
$4 = "one" OR $4 = "two"
```
(I'm not 100% sure if you can repeat parameters in Postgres, but I guess you can.) The latter version doesn't need to send the same param over the wire multiple times. Probably shouldn't make a difference unless we pass some large geometries or sth. But it should be easy enough to fix, so why not do it.

kszafran commented on provider/postgis/util_internal_test.go:
```diff
@@ -74,6 +74,111 @@ func TestReplaceTokens(t *testing.T) {
+			expectedSql:  "SELECT * FROM table ",
+			expectedArgs: []interface{}{},
+		},
+	}
```
Please add some tests for the case where the SQL has more than one `?` placeholder, since you seem to be allowing and handling such cases.

kszafran commented on server/handle_map_layer_zxy.go:
```diff
@@ -183,10 +194,58 @@ func (req HandleMapLayerZXY) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+				}
+				params[param.Token] = provider.QueryParameter{
+					Token: param.Type,
+					SQL:   "?",
```
I don't think this is correct. Shouldn't we be using `param.SQL`?

bemyak replied:
No, this branch handles the case when:
* No parameter was passed in a query
* There is a `default_value` configured

In this case we should just use the default value, so SQL becomes just `?`

kszafran replied:
Well, no. We should use the default_value instead of the value that would be passed in the query param. The SQL shouldn't change. Otherwise, you'll break the SQL.

kszafran replied:
Also, if that were the intended behavior, it wouldn't make sense to have `default_value`, because you could just say `default_sql: '7'` and it would be the same thing.

bemyak replied:
Indeed, thanks for the catch!

kszafran reviewed (commented):
Looking good so far!

bemyak committed:
feat: check all sql tokens exist

- Fixed test in case no env vars are defined
- QueryParams inserted only once per token
- Minor fixes

kszafran commented on .github/actions/amazon-linux-build-action/Dockerfile:
```diff
@@ -5,13 +5,13 @@ FROM amazonlinux:latest
-	&& echo "$GOLANG_VERSION_SHA256 golang.tar.gz" | sha256sum --strict --check \
-	&& tar -C /usr/local -xzf golang.tar.gz \
-	&& rm golang.tar.gz
+RUN curl -O https://dl.google.com/go/go$GOLANG_VERSION.linux-amd64.tar.gz \
```
Is this better? This way you have to repeat `go$GOLANG_VERSION.linux-amd64.tar.gz` everywhere instead of using the shorter `golang.tar.gz`.

bemyak replied:
As the commit comment says, I had to do this, because renaming messes up sha256 sum, so it is different from the one published on [the official website](https://go.dev/dl/). I could of course download the package, rename it and calculate the new hash, but I don't think it is the way to go.

kszafran commented on Dockerfile:
```diff
@@ -42,8 +42,8 @@ ENV BUILD_PKG="${BUILDPKG}"
-	&& apk add musl-dev=1.1.24-r10 \
-	&& apk add gcc=9.3.0-r2
+	&& apk add musl-dev \
+	&& apk add gcc
```
You could upgrade these, but I'd keep explicit versions. Otherwise you can't have reproducible builds.

kszafran replied:
BTW, I wonder why these are even needed? Does Tegola have any C dependencies?

bemyak replied:
Yep, `GpkgProvider` uses CGO

kszafran commented on Dockerfile:
```diff
@@ -61,7 +61,7 @@ RUN cd /go/src/github.com/go-spatial/tegola/cmd/tegola \
-FROM alpine:3.12
+FROM alpine:3.16
 
 RUN apk update \
```
This is out of scope of this PR, but it would be a good idea to not use the root user. E.g. https://github.com/ory/kratos/blob/master/.docker/Dockerfile-alpine#L32

bemyak replied:
Yeah, I agree that all of the CI stuff is out scope, but after running it I realized that I was relying on some features from Go 1.17 IIRC, so I decided to upgrade Go and it lead to upgrading alpine and stuff.

I don't want to make any additional changes tbh, better refactor it later.

kszafran commented on atlas/map.go:
```diff
@@ -179,7 +182,7 @@ func (m Map) FilterLayersByName(names ...string) Map {
 }
 
-func (m Map) encodeMVTProviderTile(ctx context.Context, tile *slippy.Tile) ([]byte, error) {
+func (m Map) encodeMVTProviderTile(ctx context.Context, tile *slippy.Tile, params map[string]provider.QueryParameterValue) ([]byte, error) {
```
Maybe we could have a type for this, since it's a pretty long type name 🤔. E.g.:
```go
type Params map[string]provider.QueryParameterValue
```
Your call.

kszafran commented on config/config.go:
```diff
@@ -50,55 +79,79 @@ type Webserver struct {
+		}
+
+		if _, ok := usedNames[param.Name]; ok {
+			return ErrParamNameDuplicate{
```
```suggestion
			return ErrParamDuplicateName{
```

kszafran commented on config/config.go:
```diff
@@ -50,55 +79,79 @@ type Webserver struct {
+		}
+
+		if _, ok := usedTokens[param.Token]; ok {
+			return ErrParamTokenDuplicate{
```
```suggestion
			return ErrParamDuplicateToken{
```

kszafran commented on config/config.go:
```diff
@@ -50,55 +79,79 @@ type Webserver struct {
-	// We use a negative in the name so the default is to clean
-	DontClean env.Bool `toml:"dont_clean"`
-}
+	usedNames := make(map[string]struct{})
```
Honestly, I'd probably use `map[string]bool` here for clarity, since it's just initialization code and won't have any performance impact. But if you see `struct{}` used for sets elsewhere in the existing codebase, then let's keep it for consistency.

bemyak replied:
They do use `struct{}` e.g., [here](https://github.com/go-spatial/tegola/blob/master/internal/build/tags/tags.go#L124).

I strongly prefer `struct{}` as well, since this approach doesn't have invalid invariants  on type level i.e., `false` as a value in map. I just want a set of strings, any additional data is redundant.

kszafran commented on config/config_test.go:
```diff
@@ -169,7 +170,26 @@ func TestParse(t *testing.T) {
+					token = "!PARAM2!"
+					type = "int"
+					sql = "AND ANSWER = ?"
+					default_value = "42"
```
Hm... what if we just called it `default`? And then `default_sql` would have a longer name with a qualifier, because it would probably be used way less often.

bemyak replied:
Mmmmh, I prefer to be verbose in such cases. If somebody will be wondering what does this "default" mean, then we have failed the job of designing a concise API :) Reading happens more times than writing, so I wouldn't save characters here.

Anyway, this will probably change a lot after the upstream review.

kszafran commented on config/config_test.go:
```diff
@@ -169,7 +170,26 @@ func TestParse(t *testing.T) {
+					name = "param2"
+					token = "!PARAM2!"
+					type = "int"
+					sql = "AND ANSWER = ?"
```
Let's keep column and table names lowercase for clarity. Existing test cases seem to do that, too.

kszafran commented on config/config_test.go:
```diff
@@ -1074,6 +1116,191 @@ func TestValidate(t *testing.T) {
 				},
 			},
 		},
+		"13 reserved parameter token": {
```
```suggestion
		"13 reserved token name": {
```

kszafran commented on config/config_test.go:
```diff
@@ -1074,6 +1116,191 @@ func TestValidate(t *testing.T) {
+				},
+			},
+		},
+		"13 duplicate parameter token": {
```
```suggestion
		"13 duplicate token name": {
```

kszafran commented on config/errors.go:
```diff
@@ -13,6 +15,83 @@ func (e ErrMapNotFound) Error() string {
+}
+
+func (e ErrParamNameDuplicate) Error() string {
+	return fmt.Sprintf("config: map %s redeclares duplicate parameter with name %s",
```
Either "it redeclares a parameter" or "there is a duplicate parameter" ;). Maybe:
```suggestion
	return fmt.Sprintf("config: map %s parameter %s has a name already used by another parameter",
```

kszafran commented on config/errors.go:
```diff
@@ -13,6 +15,83 @@ func (e ErrMapNotFound) Error() string {
+}
+
+func (e ErrParamTokenReserved) Error() string {
+	return fmt.Sprintf("config: map %s has parameter %s referencing reserved token %s",
```
I appreciate that you're trying to follow the format of `"map %s something"`, but I find these error messages difficult to understand. I'll suggest some alternatives.
```suggestion
	return fmt.Sprintf("config: map %s parameter %s uses a reserved token name %s",
```

kszafran commented on config/errors.go:
```diff
@@ -13,6 +15,83 @@ func (e ErrMapNotFound) Error() string {
+}
+
+func (e ErrParamTokenDuplicate) Error() string {
+	return fmt.Sprintf("config: map %s redeclares existing parameter token %s in param %s",
```
```suggestion
	return fmt.Sprintf("config: map %s parameter %s has a token name %s already used by another parameter",
```
Also, let's be consistent - use either "parameter" or "param" (no preference here on my end).

kszafran commented on config/errors.go:
```diff
@@ -13,6 +15,83 @@ func (e ErrMapNotFound) Error() string {
+	i := 0
+	for k := range provider.ParamTypeDecoders {
+		validTypes[i] = k
+		i++
```
No need for keeping a counter. Declare the slice as:
```go
validTypes := make([]string, 0, len(provider.ParamTypeDecoders))
```
and then use the good ol' `append`.

bemyak replied:
This is a neat trick :) I think you've already shown it to me, but I forgot about it.

:+1:

kszafran commented on config/errors.go:
```diff
@@ -13,6 +15,83 @@ func (e ErrMapNotFound) Error() string {
+		i++
+	}
+
+	return fmt.Sprintf("config: map %s has type %s in param %s, which is not one of the known types: %s",
```
```suggestion
	return fmt.Sprintf("config: map %s parameter %s has an unknown type, must be one of: %s",
```

kszafran commented on config/errors.go:
```diff
@@ -13,6 +15,83 @@ func (e ErrMapNotFound) Error() string {
+}
+
+func (e ErrParamTwoDefaults) Error() string {
+	return fmt.Sprintf("config: map %s has both default_value and default_sql defined in param %s",
```
```suggestion
	return fmt.Sprintf("config: map %s parameter %s has both default_value and default_sql",
```

kszafran commented on config/errors.go:
```diff
@@ -13,6 +15,83 @@ func (e ErrMapNotFound) Error() string {
+}
+
+func (e ErrParamInvalidDefault) Error() string {
+	return fmt.Sprintf("config: map %s has default value in param %s that doesn't match the parameter's type %s",
```
```suggestion
	return fmt.Sprintf("config: map %s parameter %s has a default value that is invalid for type %s",
```

kszafran commented on config/errors.go:
```diff
@@ -13,6 +15,83 @@ func (e ErrMapNotFound) Error() string {
+}
+
+func (e ErrParamBadTokenName) Error() string {
+	return fmt.Sprintf("config: map %s has parameter %s referencing token with an invalid name %s",
```
```suggestion
	return fmt.Sprintf("config: map %s parameter %s has an invalid token name %s",
```

kszafran commented on config/config_test.go:
```diff
@@ -169,7 +170,26 @@ func TestParse(t *testing.T) {
+					
+					[[maps.params]]
+					name = "param2"
+					token = "!PARAM2!"
```
I just thought... Do we even need to have a separate, configurable `token` name? Why not just use the parameter name? So if the parameter name is `favorite_food`, you'd use `!FAVORITE_FOOD!` as the token. As long as the param/token name can include a few special characters like `-_.` we should be good. I see little value in defining the parameter name and token name separately.

bemyak replied:
We could do that, however, the user will need to be aware of the transformation `x` → `!X!`, so they'll have to reference the documentation. Figuring out where a param is coming from also becomes more difficult, since you can't simply search for `!YOUR_PARAM!` anymore.

> Here goes my argument about `reading > writing` again.

So for now I'll leave is as is.

kszafran commented on provider/debug/debug.go:
```diff
@@ -25,26 +26,38 @@ func init() {
 	ext, srid := tile.Extent()
 
+	params := make([]string, len(queryParams))
+	i := 0
```
Same comment as before. Initialize the slice with 0 length and cap=len, and then use `append`. It's just cleaner and just as efficient AFAIK.

kszafran commented on provider/gpkg/gpkg_register.go:
```diff
@@ -159,6 +161,7 @@ func extractColDefsFromSQL(sql string) []string {
 func featureTableMetaData(gpkg *sql.DB) (map[string]featureTableDetails, error) {
 	// this query is used to read the metadata from the gpkg_contents, gpkg_geometry_columns, and
 	// sqlite_master tables for tables that store geographic features.
+	//goland:noinspection SqlResolve
```
I guess that was added by the original author? If there are no `goland:` instructions in the existing codebase, let's delete this one, too.

kszafran commented on provider/postgis/postgis.go:
```diff
@@ -453,7 +456,7 @@ func CreateProvider(config dict.Dicter, providerType string) (*Provider, error)
 	dbconfig, err := BuildDBConfig(uri.String())
 	if err != nil {
-		return nil, fmt.Errorf("Failed while building db config: %w", err)
+		return nil, fmt.Errorf("failed while building db config: %v", err)
```
I'd keep `%w` here. (I guess this was changed by the original author?)
```suggestion
		return nil, fmt.Errorf("failed while building db config: %w", err)
```

kszafran commented on provider/postgis/postgis.go:
```diff
@@ -472,7 +475,7 @@ func CreateProvider(config dict.Dicter, providerType string) (*Provider, error)
 	pool, err := pgxpool.ConnectConfig(context.Background(), &p.config)
 	if err != nil {
-		return nil, fmt.Errorf("Failed while creating connection pool: %w", err)
+		return nil, fmt.Errorf("failed while creating connection pool: %v", err)
```
```suggestion
		return nil, fmt.Errorf("failed while creating connection pool: %w", err)
```

kszafran commented on provider/postgis/postgis.go:
```diff
@@ -488,7 +491,7 @@ func CreateProvider(config dict.Dicter, providerType string) (*Provider, error)
 		lName, err := layer.String(ConfigKeyLayerName, nil)
 		if err != nil {
-			return nil, fmt.Errorf("For layer (%v) we got the following error trying to get the layer's name field: %w", i, err)
+			return nil, fmt.Errorf("for layer (%v) we got the following error trying to get the layer's name field: %v", i, err)
```
```suggestion
			return nil, fmt.Errorf("for layer (%v) we got the following error trying to get the layer's name field: %w", i, err)
```

kszafran commented on provider/postgis/postgis.go:
```diff
@@ -725,7 +740,23 @@ func (p Provider) inspectLayerGeomType(l *Layer) error {
-	rows, err := p.pool.Query(context.Background(), sql)
+	// substitute default values to parameter
+	params := extractQueryParamValues(pname, maps, l)
+	if err != nil {
```
This `if` does not do anything. It's probably a leftover from when `extractQueryParamValues` used to return an `error`.

kszafran commented on provider/postgis/util.go:
```diff
@@ -169,30 +156,49 @@ func replaceTokens(sql string, lyr *Layer, tile provider.Tile, withBuffer bool)
 }
 
-var tokenRe = regexp.MustCompile("![a-zA-Z0-9_-]+!")
+// extractQueryParamValues will remove all parameter tokens from the query
```
The comment does not reflect what the function actually does. I guess it's a leftover from some earlier approach.

kszafran commented on provider/postgis/postgis.go:
```diff
@@ -725,7 +740,23 @@ func (p Provider) inspectLayerGeomType(l *Layer) error {
+	if provider.ParameterTokenRegexp.MatchString(sql) {
+		// remove all parameter tokens for inspection
+		// crossing our fingers that the query is still valid 🤞
+		// if not, the user will have to specify `geometry_type` in the config
```
Hm... I'm not sure if we really need this "best effort" approach. It seems to me that the following holds (correct me if I'm wrong):
- if `default_value` is defined, use it
- if `default_sql` is defined, use it (it seems that this code doesn't do that)
- if neither is defined and the parameter is optional, replace the token with an empty string
- if neither is defined and the parameter is required, return an error if the geometry type is not specified explicitly

It seems to me that only in the last case we require an explicit geometry type. In other words, this should behave the same way as when serving a request.

bemyak replied:
> if `default_sql` is defined, use it (it seems that this code doesn't do that)

It does, `*QueryParameter.ToDefaultValue` handles this.

> if neither is defined and the parameter is optional/required

If neither is defined, the parameter is required as long as its token occurs in an SQL expression. We don't have an explicit `required` parameter now, and I don't think we should. The current approach have only one invalid invariant: when you specify both `defaul_value` and `defaul_sql`. Having the `required` config option will produce much more without giving any benefits IMO.

Anyway, this does behave the same way as during request processing (we use the same `*QueryParameter.ToDefaultValue` method), except that it tries to go an extra step further: tries to remove tokens that have left after resolving built-in one and custom with defaults. This could work in some cases that I consider pretty common, e.g.,

```
SELECT … WHERE wkb_geometry && !BBOX! !PARAM!
```
```
  [[maps.params]]
  name    = "param"
  token   = "!PARAM!"
  type = "string"
  sql = "AND osm_id == ?"
```

kszafran commented on provider/postgis/util_internal_test.go:
```diff
@@ -74,6 +74,128 @@ func TestReplaceTokens(t *testing.T) {
+			expectedSql:  "SELECT * FROM table WHERE PARAM = $1",
+			expectedArgs: []interface{}{"test"},
+		},
+		"null replacement": {
```
Hm... what is the use case here? Or, when would this happen?

bemyak replied:
Mostly I just wanted to check that the `nil` value will be handled correctly in case we'll extend it with more types in the future. But you're right, this shouldn't happen normally.

kszafran commented on provider/postgis/util_internal_test.go:
```diff
@@ -74,6 +74,128 @@ func TestReplaceTokens(t *testing.T) {
+			params: map[string]provider.QueryParameterValue{
+				"!PARAM!": {
+					Token: "!PARAM!",
+					SQL:   "",
```
Hm... is this the case where we set `default_sql` to an empty string?

bemyak replied:
Yes, it means the token must be omitted if no value is passed

kszafran commented on provider/postgis/util_internal_test.go:
```diff
@@ -74,6 +74,128 @@ func TestReplaceTokens(t *testing.T) {
+			params: map[string]provider.QueryParameterValue{
+				"!PARAM!": {
+					Token: "!PARAM!",
+					SQL:   "???",
```
Let's maybe use spaces, to at least make it look kind of correct:
```suggestion
					SQL:   "? ? ?",
```

kszafran commented on provider/provider.go:
```diff
@@ -104,13 +105,16 @@ type Tile interface {
 }
 
+// ParameterTokenRegexp to validate QueryParameters
+var ParameterTokenRegexp = regexp.MustCompile("![a-zA-Z0-9_-]+!")
```
If we go with my suggestion of using the param name as the token name, Then I'd allow `.`, too:
```suggestion
var ParameterTokenRegexp = regexp.MustCompile("![a-zA-Z0-9._-]+!")
```

kszafran commented on provider/query_parameter.go:
```diff
@@ -0,0 +1,58 @@
+	// DefaultSQL replaces SQL if param wasn't passed. Either default_sql or
+	// default_value can be specified
+	DefaultSQL   string `toml:"default_sql"`
+	DefaultValue string `toml:"default_value"`
```
What do you think about supporting `required`? Or should we assume that query parameters are necessarily optional? I feel like supporting `required` is simple enough and might be useful in some cases.

bemyak replied:
See my comment [here](https://github.com/SharperShape/tegola/pull/6#discussion_r898018121)

kszafran commented on provider/query_parameter.go:
```diff
@@ -0,0 +1,58 @@
+			SQL:   param.DefaultSQL,
+			Value: nil,
+		}, nil
+	} else {
```
Hm... IMO we should support two cases:
- Parameter doesn't have `default_value` or `default_sql`, but it not marked as `required`. In that case we replace the token with an empty string.
- Like above, but the parameter is required. Then we can return this error (and make sure it ends up as 400).

kszafran commented on provider/query_parameter_value.go:
```diff
@@ -0,0 +1,52 @@
+	// Value that will be passed to the final query in arguments list
+	Value interface{}
+	// Raw parameter values for debugging and monitoring
+	RawValues map[string]string
```
Hm... why is this a map and not just a string? When does it end up with multiple raw values?

bemyak replied:
Hm, it should be a tuple `(string, string)`, but since Go doesn't have tuples in std, I'll split this field in two: `RawParam` and `RawValue`

kszafran commented on provider/query_parameter_value.go:
```diff
@@ -0,0 +1,52 @@
+		return sql
+	}
+
+	for _, token := range ParameterTokenRegexp.FindAllString(sql, -1) {
```
I assume that at the point when `ReplaceParams` is called, all built-in tokens have already been replaced?

bemyak replied:
Yes, but shouldn't matter. After finding a token in the incoming SQL we look for it in the list of parameter tokens and then replace. I will add a check to ensure that the token was indeed found.

kszafran commented on provider/query_parameter_value.go:
```diff
@@ -0,0 +1,52 @@
+		}
+
+		// Finally, replace current token with the prepared SQL
+		sql = strings.Replace(sql, token, paramSQL, 1)
```
I'd suggest:
```suggestion
		sql = strings.ReplaceAll(sql, token, paramSQL)
```
And in the beginning of the loop check if a token has already been encountered. This way we won't have to pass the same arg multiple times if the token is used more than once.

kszafran commented on provider/query_parameter_value.go:
```diff
@@ -0,0 +1,52 @@
+		param := params[token]
+
+		// Replace every ? in the param's SQL with a positional argument
+		paramSQL := ""
```
How about using a `strings.Builder` since we build this string char by char?

Better yet, couldn't this just be done with a regular `strings.ReplaceAll`? I assume that `param.SQL` contains at least one `?`. If it doesn't, that should be a configuration error (we should validate that `sql` and `default_sql` contain at least one `?`). And then all you really need is:
```go
*args = append(*args, param.Value)
paramSQL := strings.ReplaceAll(paramSQL, "?", fmt.Sprintf("$%d", len(*args)))
```

bemyak replied:
A `default_sql` value is not allowed to have `?`. I wouldn't require it for `sql` either to support use-cases when a part of a query may or may not be added based on a flag presence.

So we don't know if we should append to `args` before the replacement has happened. Unfortunately, `strings.ReplaceAll` doesn't tell us if it actually did something or not, so in this case we'll have to compare strings before and after the replacement, or check if `?` occurs before that.

This will require additional iteration, so I'd prefer to avoid it.

kszafran replied:
Hm... I didn't think about using params as flags 🤔. I'm not sure if that's a very common use case. And even then, you should be able to do it even if you require `sql` to have `?` (`default_sql` should actually be required to NOT have `?`, my bad). You could have e.g. `sql: "...whatever... AND ? = ?"`. I'd make `?` mandatory in `sql` and keep it simple.

bemyak replied:
Anyway, the point is that in this method we don't know if `param.SQL` was an `sql` or `default_(value|sql)`, so we can't rely on `?` presence (since `defefault_sql` doesn't have it). It means iterating twice through the string(s). I'd prefer to just do replacement "by hand", not a big deal.

kszafran replied:
Alrighty, let's just use `strings.Builder`, then.

kszafran commented on server/handle_map_layer_zxy.go:
```diff
@@ -183,10 +193,42 @@ func (req HandleMapLayerZXY) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	_, err = w.Write(pbyte)
+	if err != nil {
+		log.Errorf("error writing tile z:%v, x:%v, y:%v - %v", req.z, req.x, req.y, err)
```
Hm... Let's see what maintainers think about it, but we might want to log this on another level. It think this could happen e.g. when the client closes the connection or cancels the request and that's extremely common. If you move the map before a tile loads, the request will probably be canceled. This way we may end up spamming the logs a lot.

bemyak replied:
This check was added by the original PR author, previously it was just ignored. I'll change it to an Info-level message.

kszafran reviewed (changes requested):
Good progress! I left some more suggestions.

Also, before making a PR to the upstream, let's try to use this version in our service and see if it actually works and fulfills our requirements.

dependabot[bot] committed:
Bump shell-quote from 1.7.2 to 1.7.3 in /ui

Bumps [shell-quote](https://github.com/substack/node-shell-quote) from 1.7.2 to 1.7.3.
- [Release notes](https://github.com/substack/node-shell-quote/releases)
- [Changelog](https://github.com/substack/node-shell-quote/blob/master/CHANGELOG.md)
- [Commits](https://github.com/substack/node-shell-quote/compare/v1.7.2...1.7.3)

---
updated-dependencies:
- dependency-name: shell-quote
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

iwpnd committed:
fix: filter by zoom level for min/max zoom 0

Prior to this, setting min_zoom=0 and max_zoom=0 would result
in the layer being present across the entire zoom range.

dependabot[bot] committed:
Bump terser from 4.8.0 to 4.8.1 in /ui

Bumps [terser](https://github.com/terser/terser) from 4.8.0 to 4.8.1.
- [Release notes](https://github.com/terser/terser/releases)
- [Changelog](https://github.com/terser/terser/blob/master/CHANGELOG.md)
- [Commits](https://github.com/terser/terser/commits)

---
updated-dependencies:
- dependency-name: terser
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

iwpnd committed:
fix: config max zoom 0

as per #870, setting max_zoom to 0 currently causes the layer to bleed through
over the entire zoom range. Either we fix filtering as per #871
or make sure max_zoom is never 0 as I do in this PR.

roelarents committed:
handle gpkg GEOMETRY as nil

Commit:
Added extra parameter to seed command to allow users to control which tiles produce logging output

Juneezee committed:
test: use `T.Setenv` to set env vars in tests

This commit replaces `os.Setenv` with `t.Setenv` in tests. The
environment variable is automatically restored to its original value
when the test and all its subtests complete.

Reference: https://pkg.go.dev/testing#T.Setenv
Signed-off-by: Eng Zer Jun <engzerjun@gmail.com>

bemyak committed:
feat: disable caching for maps with params

Warn about disabled caching

bemyak committed:
refactor: move replaceParams fn to provider

bemyak committed:
refactor: move Map and related types to provider pkg

This is needed to access configured map parameters from the provider's
initialization code to inspect geom type. Unfortunately, config pkg
depends on the provider pkg, so simply referencing these types creates a
dependency loop.

bemyak committed:
feat: disable caching for maps with custom params

ARolek committed:
README updates

ARolek committed:
Bump Go version

ARolek committed:
Bump Alpine 3.16 deps

bemyak commented:
Closing; merged into upstream.
